hack/optimize/neuron/Dockerfile

```diff
@@ -0,0 +1,73 @@
+# Use Ubuntu 20.04 as the base image
+FROM ubuntu:20.04
+
+# Neuron SDK components versions
+ARG NEURONX_FRAMEWORK_VERSION=2.11.0.0
+ARG NEURONX_RUNTIME_LIB_VERSION=2.11.7.0
+ARG NEURONX_TOOLS_VERSION=2.11.8.0
+ARG NEURONX_CC_VERSION=2.11.8.0
+
+# Set environment variables
+ENV DEBIAN_FRONTEND=noninteractive
+ENV PYTHONUNBUFFERED=1
+ENV PYTHONIOENCODING=UTF-8
+ENV LD_LIBRARY_PATH="/opt/aws/neuron/lib:/usr/local/lib"
+ENV PATH="/opt/aws/neuron/bin:$PATH"
+
+# Install system dependencies including libsqlite3-dev and libbz2-dev for Python
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+    build-essential \
+    ca-certificates \
+    curl \
+    wget \
+    zlib1g-dev \
+    gnupg2 \
+    libssl-dev \
+    libffi-dev \
+    libsqlite3-dev \
+    libbz2-dev \
+    libopenblas-dev \
+    libomp5 \
+    && rm -rf /var/lib/apt/lists/*
+
+# Add Neuron repository and install Neuron SDK components
+RUN echo "deb https://apt.repos.neuron.amazonaws.com focal main" > /etc/apt/sources.list.d/neuron.list && \
+    wget -qO - https://apt.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-NEURON.PUB | apt-key add - && \
+    apt-get update && \
+    apt-get install -y \
+    aws-neuronx-tools=${NEURONX_TOOLS_VERSION} \
+    aws-neuronx-runtime-lib=${NEURONX_RUNTIME_LIB_VERSION} \
+    && rm -rf /var/lib/apt/lists/*
+
+# Install Python 3.10 with sqlite3 and bz2 support
+RUN wget -q https://www.python.org/ftp/python/3.10.12/Python-3.10.12.tgz && \
+    tar -xzf Python-3.10.12.tgz && \
+    cd Python-3.10.12 && \
+    ./configure --enable-shared --enable-optimizations --with-ensurepip=install && \
+    make -j $(nproc) && make install && \
+    cd .. && rm -rf Python-3.10.12*
+
+# Upgrade pip and install required Python packages
+RUN python3.10 -m pip install --upgrade pip
+
+# Install Neuron-related Python packages from the Neuron repository
+RUN python3.10 -m pip install --no-cache-dir \
+    --extra-index-url https://pip.repos.neuron.amazonaws.com \
+    torch-neuronx==${NEURONX_FRAMEWORK_VERSION} \
+    torch-xla==1.13.* \
+    torchvision
+
+# Install additional Python packages
+RUN python3.10 -m pip install --no-cache-dir \
+    transformers==4.29 \
+    numpy==1.23 \
+    pynvml
+
+# Set the working directory
+WORKDIR /app
+
+# Copy training and inference scripts
+COPY train_bert_neuron.py /app/train_bert_neuron.py
+COPY infer_bert_neuron.py /app/infer_bert_neuron.py
```

Comment:
so this image supports inference and training for neuron? should we just put it under images rather than hack? 

these python scripts you could leave in `/hack` and then just volume mount them into the container

Reply:
Correct. Yeah honestly I struggled with where to put these, and someone recommended hack a couple weeks ago. The main use case right now is to just get optimal batch size to support upcoming benchmarking efforts for our e2e tests. 

I could see it evolving in the future to being automatically ran when certain dependencies are updated, or as new instance types become available.

Reply:
so IIUC we can use the neuron test for inference tuning but you need an imaage for neuron here that supports training as well?

Reply:
@ndbaker1 Use of `Dockerfile` was just to make things more portable across instances as I did testing. Also, while probably made no difference, there is slight overhead introduced from running in a container versus just a script. Additional dependencies (e.g. neuron container runtime) as well, which makes the optimization's environment closer to the tests' runtime environment.

Reply:
@ndbaker1 @cartermckinnon Not sure I have a perfect answer of where these scripts/dockerfile should go, but here's the full context...

* The training and inference tests part of e2e2 currently have suboptimal values for their batch parameter. 
* A standard batch value is hardcoded for all of them, leaving many of the instance's GPUs underutilized.
* A major goal moving forward is to be able to benchmark these tests on all instances, and to gain an understanding of what full performance optimization looks like. 

* These new optimization scripts look to target a single GPU on an instance (even if multiple GPU), and to determine max batch size that a GPU of a certain type can handle. 
* The optimal batch value will then be used to determine the total batch size per instance (batch_size * num_gpus) for each instance, enabling us to run benchmarking for each instance at full GPU utilization (like our customers would)
* The need for a training and inference script has to do with the fact that depending on the "mode" of a model, more/less memory might be utilized
* Memory utilization by mode differs because training requires large amounts of temporary parameter values to be held in memory, while inference does not
* The scripts were containerized to more closely mirror the test's runtime environment of running on kubernetes
* A single Dockerfile was used for simplicity

Reply:
Makes sense. Can we include this script in our existing test images so we don't need a separate pipeline for it? will be easier to set up a periodic for this as well if it's all the same spec with a different command

Reply:
I like this, dependencies should be kept the consistent anyways. Can't do this for Neuron yet, I'm just now noticing that the PR for Neuron BERT training/inference was closed and never merged. Will need to get that merged in first.